Move label obstacle settings to "placement" tab and show detailed settings in a stacked panel widget

nyalldawson opened the pull request:
Because:

1. The settings belong next to the placement settings, because they
affect label placement directly (and are closely associated with the
label priority setting in this tab)

2. The label settings widget is ridiculously complex and overwhelming
even for experienced QGIS users. By moving detailed settings into
stacked panels we can avoid the initial complexity of the label settings
whilst also allowing us more flexibility to expose additional settings
in future without adding to the mess. So in future, I propose we move
all detailed settings to sub panels, and leave only the initial "enable"
setting at the top level (e.g. [x] "Repeat levels" -> click settings -> see
all repeat settings).



![Peek 2019-12-04 09-58](https://user-images.githubusercontent.com/1829991/70100184-b5cba780-167c-11ea-8e29-e18afbd003aa.gif)

nyalldawson committed:
[api] New class QgsLabelSettingsWidgetBase

Base class for widgets which allow customisation of label engine properties,
such as label placement settings.

nyalldawson committed:
Hookup data defined properties support for QgsLabelSettingsWidgetBase

nyalldawson committed:
Add QgsLabelObstacleSettingsWidget widget for customising label obstacle settings

nyalldawson committed:
Add some unit tests for label settings widgets

nyalldawson committed:
[needs-docs] Move label obstacle settings to "placement" tab
and show detailed settings in a stacked panel widget

Because:

1. The settings belong next to the placement settings, because they
affect label placement directly (and are closely associated with the
label priority setting in this tab)

2. The label settings widget is ridiculously complex and overwhelming
even for experienced QGIS users. By moving detailed settings into
stacked panels we can avoid the initial complexity of the label settings
whilst also allowing us more flexibility to expose additional settings
in future without adding to the mess. So in future, I propose we move
all detailed settings to sub panels, and leave only the initial "enable"
setting at the top level (e.g. [x] "Repeat levels" -> click settings -> see
all repeat settings).

nyalldawson committed:
Fix build

nyalldawson committed:
Add a helpful label explaining obstacle weight

nirvn commented on src/ui/qgslabelobstaclesettingswidgetbase.ui:
```diff
@@ -0,0 +1,133 @@
+      <item>
+       <widget class="QLabel" name="label">
+        <property name="text">
+         <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-weight:600;&quot;&gt;Setting a higher weight for obstacles means that labels are less likely to cover features in this layer.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;Only labels with a higher priority then this obstacle weight will be placed over these obstacles.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
More verbose description most welcome.

I'm wondering whether part of this description should be conditional to the labelling engine version selected.

nyalldawson committed:
Fix spelling

nyalldawson committed:
Avoid fragile combo box index based logic

nyalldawson committed:
[labeling] When a layer is set to "blocking" mode, expose the obstacle
settings widget to allow users to tweak the obstacle behavior

Previously these settings were stuck inside the disabled label settings
widget, which meant that to edit them you had to temporarily enable simple
labels, tweak the setting, and then set back to obstacle mode... woot.

nyalldawson commented:
Bonus fix: we can now actually change the obstacle settings when a layer is set to blocking mode (Previously these settings were stuck inside the disabled label settings widget, which meant that to edit them you had to temporarily enable simple labels, tweak the setting, and then set back to obstacle mode... woot.)

![image](https://user-images.githubusercontent.com/1829991/70106255-d1d84480-168e-11ea-9432-f96aade7e728.png)

NathanW2 commented:
On this note: How do we feel about moving the placement tab to be the
second tab?  I don't like it's a primary thing you normally do but is so
far away from the user.

On Wed, Dec 4, 2019 at 12:09 PM Nyall Dawson <notifications@github.com>
wrote:

> Bonus fix: we can now actually change the obstacle settings when a layer
> is set to blocking mode (Previously these settings were stuck inside the
> disabled label settings widget, which meant that to edit them you had to
> temporarily enable simple labels, tweak the setting, and then set back to
> obstacle mode... woot.)
>
> [image: image]
> <https://user-images.githubusercontent.com/1829991/70106255-d1d84480-168e-11ea-9432-f96aade7e728.png>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/33214?email_source=notifications&email_token=AAC5FXHRK4WMGLVKNAUZEOTQW4GNBA5CNFSM4JVAPJQKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEF3PAOY#issuecomment-561442875>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC5FXFUECFBTL6K7OCV3C3QW4GNBANCNFSM4JVAPJQA>
> .
>

nirvn commented:
@NathanW2 , I'm not super warm to the idea of moving the tab.

NathanW2 commented:
Why is that?

On Wed, Dec 4, 2019 at 1:19 PM Mathieu Pellerin <notifications@github.com>
wrote:

> @NathanW2 <https://github.com/NathanW2> , I'm not super warm to the idea
> of moving the tab.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/33214?email_source=notifications&email_token=AAC5FXFX7L4G5EIPYC5BRJDQW4OTFA5CNFSM4JVAPJQKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEF3SZGQ#issuecomment-561458330>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC5FXATCFFJG3VTILH242DQW4OTFANCNFSM4JVAPJQA>
> .
>

nirvn commented:
@NathanW2 , mainly because this tab isn't always shown in the overall text formatting widget (i.e., outside of labels, the last tabs are hidden). I think it's more effective to have the tabs that always show first, helping muscle memory, etc. And yes, on that front, it does ultimately mean we should move the masking tab.

Also, I feel the text tab and formatting tab are so closely linked to one another it would be a shame to break those brothers apart ;)
